### PR TITLE
Support warp shuffle and sync functions in CuPy JIT

### PR DESCRIPTION
**Pull request by leofang**
Part of #5280.

Example:
```python
import cupy as cp
from cupyx import jit

@jit.rawkernel()
def f_shfl(a, b):
    laneId = jit.threadIdx.x & 0x1f
    if laneId == 0:
        value = a
    value = jit.shfl_sync(0xffffffff, value, 0)
    b[laneId] = value


a = cp.int64(100)
b = cp.empty((32,), dtype=cp.int64)
f_shfl[1, 32](a, b)
print(b)


@jit.rawkernel()
def f_shfl_up(a):
    value = jit.shfl_up_sync(0xffffffff, a[jit.threadIdx.x], 5)
    a[jit.threadIdx.x] = value

a = cp.arange(32, dtype=cp.int64)
f_shfl_up[1, 32](a)
print(a)
```
Output:
```
[100 100 100 100 100 100 100 100 100 100 100 100 100 100 100 100 100 100
 100 100 100 100 100 100 100 100 100 100 100 100 100 100]
[ 0  1  2  3  4  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18
 19 20 21 22 23 24 25 26]
```

TODO:
- [x] Support `__syncwarp()`
- [x] Add docs
- [x] Add tests

### CODE REVIEW

File: cupyx/jit/_builtin_funcs.py [hunk 1]
```diff
@@ -94,14 +96,53 @@ def call(self, env, *args, **kwds):
...
         """
         super.__call__(self)
 
     def call_const(self, env):
         return Data('__syncthreads()', _cuda_types.void)
 
 
+class SyncWarp(BuiltinFunc):
+
+    def __call__(self, mask=0xffffffff):
```
Comment (asi1024) [inline cupyx/jit/_builtin_funcs.py:114]: Could you fix `mask` to keyword only arugment?
```suggestion
    def __call__(self, *, mask=0xffffffff):
```

File: cupyx/jit/_builtin_funcs.py [hunk 2]
```diff
@@ -94,14 +96,53 @@ def call(self, env, *args, **kwds):
...
+            mask (int): Active threads in a warp. Default is 0xffffffff.
+
+        .. seealso:: `Synchronization functions`_
+
+        .. _Synchronization functions:
+            https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#synchronization-functions
+        """
+        super().__call__()
+
+    def call(self, env, **kwargs):
```
Comment (asi1024) [inline cupyx/jit/_builtin_funcs.py:127]: How about writing in this style?
```suggestion
    def call(self, env, *, mask=None):
```

File: cupyx/jit/_builtin_funcs.py (hunk 1 above)
Comment (leofang) [reply cupyx/jit/_builtin_funcs.py:114]: Yes `mask` is meant to be a kw-only argument. Thanks @asi1024!

**Comment by asi1024** [comment]
Jenkins, test this please.

**Review by asi1024 — APPROVED** [review]

**Comment by asi1024** [comment]
LGTM. Thanks!

**Pull request merged**